hilti/toolchain/include/compiler/validator.h

```diff
@@ -42,9 +43,23 @@ class VisitorMixIn {
     void checkTypeArguments(const node::Range<Expression>& have, const node::Set<type::function::Parameter>& want,
                             Node* n, bool allow_no_arguments = false, bool do_not_check_types = false);
 
+    /**
+     * Ensures that the node represented by tag can have all of the provided
+     * attributes. This does not use any context, if more information is needed,
+     * then do the check elsewhere.
+     */
+    void checkNodeAttributes(node::Tag tag, AttributeSet* attributes, const std::string_view& where);
+
 private:
     Builder* _builder;
     int _errors = 0;
+
+    static inline std::unordered_map<node::Tag, std::unordered_set<Attribute::Kind>> allowed_attrs{
```

Comment:
This feels like a bit of an awkward location for the table. How about moving this to the top of `validator.cc` as a static global (not class member)?